File: SatelliteAndroid/satellite/src/main/java/yuni/library/satellite/LocationManager.java
package yuni.library.satellite;

import android.util.Log;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

class LocationManager {

    private LocationService mLocationService;
    private Location mLastLocation;

    private List<ContinueLocationHolder> mHolderList = new ArrayList<>();
    private Satellite.Options mOptions;

    public LocationManager(LocationService locationService) {
        mLocationService = locationService;
        mLocationService.setListener(new Satellite.Listener() {
            @Override
            public void onLocationChanged(Location location) {
                mLastLocation = location;
                notifyLocationChanged(location);
            }
        });
    }

    public Location getLastLocation() {
        if (mLastLocation == null) {
            return mLocationService.getLastLocation();
        }
        return mLastLocation.copy();
    }

    public Satellite.Options getUsingOptions() {
        if (mOptions == null) {
            return mOptions;
        }
        return mOptions.copy();
    }

    public synchronized ContinueLocationHolder startContinue(Object owner, Satellite.Options options, Satellite.Listener listener) {
        options = options.copy();

        ContinueLocationHolder holder = new ContinueLocationHolder(owner);
        holder.options = options;
        holder.listener = listener;
        mHolderList.add(holder);

        if (computeOptions()) {
            startOrStop();
        }

        return holder;
    }

    public synchronized void updateOptions(ContinueLocationHolder holder, Satellite.Options options) {
        options = options.copy();
        holder.options = options;
        if (computeOptions()) {
            startOrStop();
        }
    }

    public synchronized void stopContinue(ContinueLocationHolder holder) {
        if (!mHolderList.remove(holder)) {
            return;
        }

        if (computeOptions()) {
            startOrStop();
        }
    }

    public void getLocationOnce(Satellite.Options options, final Satellite.Listener listener) {
        if (mLastLocation != null && options.isCache() && !options.address) {
            long now = getCurrentTime();
            if (now - mLastLocation.getTime() < options.cacheTime) {
                listener.onLocationChanged(mLastLocation.copy());
                return;
            }
        }
        mLocationService.getOnce(options.copy(), location -> {
            mLastLocation = location;
            listener.onLocationChanged(location);
        });
    }

    public void checkOwners() {
        if (computeOptions()) {
            startOrStop();
        }
    }

    long getCurrentTime() {
        return System.currentTimeMillis();
    }

    private synchronized void notifyLocationChanged(Location location) {
        mLastLocation = location;

        for (int i=mHolderList.size()-1; i>=0; i--) {
            ContinueLocationHolder holder = mHolderList.get(i);
            if (holder.ownerRef.get() == null) {
                warnWeakReleased(holder.ownerKey);
                continue;
            }
            holder.listener.onLocationChanged(location);
        }

        if (computeOptions()) {
            startOrStop();
        }
    }

    private boolean computeOptions() {
        int mode = 0;
        long interval = Long.MAX_VALUE;
        boolean address = false;

        for (int i=mHolderList.size()-1; i>=0; i--) {
            ContinueLocationHolder holder = mHolderList.get(i);

            if (holder.ownerRef.get() == null) {
                mHolderList.remove(i);
                warnWeakReleased(holder.ownerKey);
                continue;
            }

            Satellite.Options options = holder.options;

            // 模式计算
            if (options.mode > mode) {
                mode = options.mode;
            }

            // interval
            if (options.interval < interval ) {
                interval = options.interval;
            }

            // address
            address |= options.address;
        }

        if (mHolderList.isEmpty()) {
            return true;
        }

        boolean changed = false;
        if (mOptions == null) {
            mOptions = new Satellite.Options();
            changed = true;
        }

        if (mOptions.mode != mode) {
            changed = true;
            mOptions.mode = mode;
        }

        if (mOptions.interval != interval) {
            changed = true;
            mOptions.interval = interval;
        }

        if (mOptions.address != address) {
            changed = true;
            mOptions.address = address;
        }

        if (!mLocationService.isStarted()) {
            return true;
        }

        return changed;
    }

    private void startOrStop() {
        if (mHolderList.isEmpty()) {
            mLocationService.stop();
            mOptions = null;
            return;
        }
        if (mLocationService.isStarted()) {
            mLocationService.updateOptions(mOptions);
        } else {
            mLocationService.start(mOptions);
        }
    }

    private void warnWeakReleased(String key) {
        Satellite.shared.log(Log.WARN, "LocationHandler has been released but stopContinue not called, owned by " + key);
    }

    class ContinueLocationHolder {

        WeakReference<Object> ownerRef;
        String ownerKey;
        Satellite.Options options;
        Satellite.Listener listener;

        public ContinueLocationHolder(Object owner) {
            this.ownerRef = new WeakReference<>(owner);
            this.ownerKey = owner.toString();
        }
    }

}

File: SatelliteIOS/SatelliteIOS/Classes/LocationManager.swift
//
//  LocationManager.swift
//  SatelliteIOS
//
//  Created by mingo on 2018/8/3.
//

import Foundation
import CoreLocation

class LocationManager {
    
    let locationService: LocationService
    
    var lock: NSRecursiveLock = NSRecursiveLock()
    private(set) var usingOptions: Satellite.Options?
    private var holderList: [ContinueLocationHolder] = []
    
    var lastLocation: CLLocation? {
        return _lastLocation ?? locationService.lastLocation
    }
    
    private var _lastLocation: CLLocation?
    
    init(locationService: LocationService) {
        self.locationService = locationService
        self.locationService.setListener { [weak self] (error, location) in
            guard let sself = self else { return }
            guard let location = location else { return }
            sself._lastLocation = location
            sself.notifyLocationChanged(location)
        }
    }
    
    func startContinue(owner: AnyObject, options: Satellite.Options, listener: @escaping SatelliteListener) -> ContinueLocationHolder {
        lock.lock(); defer { lock.unlock() }
        
        let holder = ContinueLocationHolder()
        holder.owner = owner
        holder.options = options
        holder.listener = listener
        holderList.append(holder)
        
        if computeOptions() {
            startOrStop()
        }
        
        return holder
    }
    
    func stopContinue(holder: ContinueLocationHolder) {
        lock.lock(); defer { lock.unlock() }
        
        var removed = false
        for i in (0...holderList.count-1) {
            if holder == holderList[i] {
                holderList.remove(at: i)
                removed = true
                break
            }
        }
        
        if !removed {
            return
        }
        
        if computeOptions() {
            startOrStop()
        }
    }
    
    func updateOptions(holder: ContinueLocationHolder, options: Satellite.Options) {
        lock.lock(); defer { lock.unlock() }
        
        holder.options = options.clone()
        if computeOptions() {
            startOrStop()
        }
    }
    
    func getLocationOnce(options: Satellite.Options, listener: @escaping SatelliteListener) {
        if let lastLocation = self.lastLocation, options.cache {
            let now = getCurrentTime()
            if now - lastLocation.timestamp.timeIntervalSince1970 < options.cacheTime {
                listener(nil, lastLocation)
                return
            }
        }
        locationService.getOnce(options: options.clone(), listener: listener)
    }
    
    func checkOwners() {
        if computeOptions() {
            startOrStop()
        }
    }
    
    func getCurrentTime() -> TimeInterval {
        return Date().timeIntervalSince1970
    }
    
    private func computeOptions() -> Bool {
        lock.lock(); defer { lock.unlock() }
        
        var distanceFilter: CLLocationDistance = 999999999
        var desiredAccuracy: CLLocationAccuracy = 999999999
        
        for i in (0..<holderList.count).reversed() {
            let holder = holderList[i]
            if holder.owner == nil {
                holderList.remove(at: i)
                continue
            }
            
            let options = holder.options!
            if options.distanceFilter < distanceFilter {
                distanceFilter = options.distanceFilter
            }
            if options.desiredAccuracy < desiredAccuracy {
                desiredAccuracy = options.desiredAccuracy
            }
        }
        
        if holderList.isEmpty {
            return true
        }
        
        var changed = false
        if self.usingOptions == nil {
            self.usingOptions = Satellite.Options()
            changed = true
        }
        
        var options = self.usingOptions!
        
        if options.desiredAccuracy != desiredAccuracy {
            options.desiredAccuracy = desiredAccuracy
            changed = true
        }
        
        if options.distanceFilter != distanceFilter {
            options.distanceFilter = distanceFilter
            changed = true
        }
        
        self.usingOptions = options
        
        if !self.locationService.isStarted {
            return true
        }
        
        return changed
    }
    
    private func startOrStop() {
        lock.lock(); defer { lock.unlock() }
        
        if holderList.isEmpty {
            self.locationService.stop()
            self.usingOptions = nil
            return
        }
        
        if self.locationService.isStarted {
            self.locationService.updateOptions(self.usingOptions!)
        } else {
            self.locationService.start(options: self.usingOptions!)
        }
    }
    
    private func notifyLocationChanged(_ location: CLLocation) {
        lock.lock(); defer { lock.unlock() }
        
        for holder in holderList {
            if holder.owner != nil {
                holder.listener(nil, location)
            }
        }
        
        if computeOptions() {
            startOrStop()
        }
    }
    
    class ContinueLocationHolder : Equatable {
        
        static func == (lhs: ContinueLocationHolder, rhs: ContinueLocationHolder) -> Bool {
            return lhs == rhs
        }
        
        weak var owner: AnyObject?
        var options: Satellite.Options!
        var listener: SatelliteListener!
    }
    
}
